lib/Handler.php

```diff
@@ -96,23 +96,31 @@ public function delete(INotification $notification): array {
 		$deleted = [];
 		$notifications = [];
 		while ($row = $statement->fetch()) {
-			if (!isset($deleted[$row['user']])) {
-				$deleted[$row['user']] = [];
-			}
-
-			$deleted[$row['user']][] = [
-				'id' => (int) $row['notification_id'],
-				'app' => $row['app'],
-			];
 			$notifications[(int) $row['notification_id']] = $this->notificationFromRow($row);
 		}
 		$statement->closeCursor();
 
-		foreach ($deleted as $user => $entries) {
-			foreach ($entries as $entry) {
-				$this->deleteById($entry['id'], (string) $user, $notifications[$entry['id']]);
+		$this->connection->beginTransaction();
+		try {
+			$shouldFlush = $this->manager->defer();
+
+			foreach ($notifications as $n) {
+				$this->manager->dismissNotification($n);
+			}
+
+			$notificationIds = array_keys($notifications);
+			foreach (array_chunk($notificationIds, 1000) as $chunk) {
+				$this->deleteIds($chunk);
+			}
+
+			if ($shouldFlush) {
+				$this->manager->flush();
 			}
+		} catch (\Throwable $e) {
+			$this->connection->rollBack();
+			throw $e;
 		}
+		$this->connection->commit();
 
 		return $deleted;
```

Comment:
$deleted will always be empty here. If intended replace by return []; and maybe add a comment.

Reply:
Right, thanks for the pointed. there is yet another loop of events going out and in again. let me see if we can optimize that

Reply:
Thanks for the hint. Fixed